### PR TITLE
import into: log checksum for each KV group

### PR DESCRIPTION
**Pull request by lance6716**
### What problem does this PR solve?

Issue Number: ref #50832

Problem Summary:

### What changed and how does it work?

- prints log for the checksum of each KV group
- move the calculation of checksum from delivering goroutine to encoding goroutine, to avoid checksum is lost after encoding before delivering
- unify the two encoders for file source and query source by using a `readFn`

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [x] Manual test: pass 40T test
- [x] No need to test
  > pass existing tests

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: br/pkg/lightning/verification/checksum.go [hunk 1]
```diff
@@ -127,3 +128,136 @@ func (c *KVChecksum) MarshalJSON() ([]byte, error) {
...
 }
+
+// KVGroupChecksum is KVChecksum(s) each for a data KV group or index KV groups.
+type KVGroupChecksum struct {
+	m     map[int64]*KVChecksum
+	codec tikv.Codec
+}
+
+// DataKVGroupAsIndexID represents the index ID for data KV group.
+const DataKVGroupAsIndexID = -1
```
Comment (D3Hunter) [inline br/pkg/lightning/verification/checksum.go:139]:
```suggestion
// DataKVGroupID represents the ID for data KV group, as index id starts from 1, so we use -1 to represent data kv group.
const DataKVGroupID = -1
```

**Review by D3Hunter — COMMENTED** [review]
will review later

File: br/pkg/lightning/verification/checksum.go [hunk 2]
```diff
@@ -127,3 +128,136 @@ func (c *KVChecksum) MarshalJSON() ([]byte, error) {
...
+	cksum := c.m[indexID]
+	if cksum == nil {
+		cksum = NewKVChecksumWithKeyspace(c.codec)
+		c.m[indexID] = cksum
+	}
+	cksum.UpdateOne(kv)
+}
+
+// Add adds the checksum of another KVGroupChecksum.
+func (c *KVGroupChecksum) Add(other *KVGroupChecksum) {
```
Comment (D3Hunter) [inline br/pkg/lightning/verification/checksum.go:176]:
```suggestion
func (c *KVGroupChecksum) Merge(other *KVGroupChecksum) {
```

File: pkg/disttask/importinto/subtask_executor.go [hunk 3]
```diff
@@ -109,13 +128,24 @@ func postProcess(ctx context.Context, taskMeta *TaskMeta, subtaskMeta *PostProce
 	// 	err = multierr.Append(err, err2)
 	// }()
 
-	localChecksum := verify.MakeKVChecksum(subtaskMeta.Checksum.Size, subtaskMeta.Checksum.KVs, subtaskMeta.Checksum.Sum)
+	localChecksum := verify.NewKVGroupChecksumForAdd()
+	for id, cksum := range subtaskMeta.Checksum {
+		callLog.Info(
+			"group checksum",
```
Comment (D3Hunter) [inline pkg/disttask/importinto/subtask_executor.go:134]:
```suggestion
			"kv group checksum",
```

File: pkg/disttask/importinto/task_executor.go [hunk 4]
```diff
@@ -224,11 +225,19 @@ func (s *importStepExecutor) OnFinished(ctx context.Context, subtask *proto.Subt
...
+	subtaskMeta.Checksum = map[int64]Checksum{}
+	for id, c := range sharedVars.Checksum.GetInnerChecksums() {
+		subtaskMeta.Checksum[id] = Checksum{
+			Sum:  c.Sum(),
+			KVs:  c.SumKVS(),
+			Size: c.SumSize(),
+		}
+		if s.tableImporter.IsGlobalSort() && id == verification.DataKVGroupAsIndexID {
+			dataKVCount = c.SumKVS()
+		}
```
Comment (D3Hunter) [inline pkg/disttask/importinto/task_executor.go:237]: see above comment, this count doesn't matter for global sort

File: pkg/executor/importer/chunk_process.go [hunk 5]
```diff
@@ -48,187 +48,260 @@ var (
...
-			kvSize += kvs.Size()
-			rowCount++
-			// pebble cannot allow > 4.0G kv in one batch.
-			// we will meet pebble panic when import sql file and each kv has the size larger than 4G / maxKvPairsCnt.
-			// so add this check.
-			if kvSize >= MinDeliverBytes || len(rowBatch) >= MinDeliverRowCnt || readPos == p.chunkInfo.Chunk.EndOffset {
-				canDeliver = true
-			}
+		if currOffset >= 0 && metrics != nil {
+			delta := currOffset - prevOffset
```
Comment (D3Hunter) [inline pkg/executor/importer/chunk_process.go:227]: `fileChunkEncoder) init()`(prevOffset) is lost, some part of the file might be accounted multiple times in this metric

**Review by D3Hunter — COMMENTED** [review]
rest lgtm

File: br/pkg/lightning/verification/checksum.go (hunk 2 above)
Comment (lance6716) [reply br/pkg/lightning/verification/checksum.go:176]: This function name is the same as `func (c *KVChecksum) Add(other *KVChecksum)`, maybe it represent checksum can have `+` operator like numbers.

File: pkg/executor/importer/chunk_process.go (hunk 5 above)
Comment (lance6716) [reply pkg/executor/importer/chunk_process.go]: https://github.com/pingcap/tidb/pull/50871/commits/0fee9cf75b6eb8450080141412d097a7db3f4200

**Review by D3Hunter — APPROVED** [review]

**Review by Benjamin2037 — APPROVED** [review]
LGTM

**Pull request merged**